src/java.desktop/share/classes/javax/swing/plaf/metal/MetalBorders.java

```diff
@@ -241,41 +263,90 @@ public void paintBorder(Component c, Graphics g, int x, int y,
                 shadow = MetalLookAndFeel.getControlInfo();
             }
 
-              g.setColor(background);
-              // Draw outermost lines
-              g.drawLine( 1, 0, w-2, 0);
-              g.drawLine( 0, 1, 0, h-2);
-              g.drawLine( w-1, 1, w-1, h-2);
-              g.drawLine( 1, h-1, w-2, h-1);
+            Graphics2D g2d = (Graphics2D) g;
+            AffineTransform at = g2d.getTransform();
+            Stroke oldStk = g2d.getStroke();
+            Color oldColor = g2d.getColor();
+            boolean resetTransform;
+            int stkWidth = 1;
+
+            // if m01 or m10 is non-zero, then there is a rotation or shear
+            // skip resetting the transform
+            resetTransform = (at.getShearX() == 0) && (at.getShearY() == 0);
+
+            if (resetTransform) {
+                g2d.setTransform(new AffineTransform());
+                stkWidth = Region.clipRound(Math.min(at.getScaleX(), at.getScaleY()));
+                g2d.setStroke(new BasicStroke((float) stkWidth));
+            }
 
-              // Draw the bulk of the border
-              for (int i = 1; i < 5; i++) {
-                  g.drawRect(x+i,y+i,w-(i*2)-1, h-(i*2)-1);
-              }
+            int xtranslation = 0;
+            int ytranslation = 0;
+            int width = 0;
+            int height = 0;
 
-              if (c instanceof JInternalFrame &&
-                               ((JInternalFrame)c).isResizable()) {
-                  g.setColor(highlight);
-                  // Draw the Long highlight lines
-                  g.drawLine( corner+1, 3, w-corner, 3);
-                  g.drawLine( 3, corner+1, 3, h-corner);
-                  g.drawLine( w-2, corner+1, w-2, h-corner);
-                  g.drawLine( corner+1, h-2, w-corner, h-2);
-
-                  g.setColor(shadow);
-                  // Draw the Long shadow lines
-                  g.drawLine( corner, 2, w-corner-1, 2);
-                  g.drawLine( 2, corner, 2, h-corner-1);
-                  g.drawLine( w-3, corner, w-3, h-corner-1);
-                  g.drawLine( corner, h-3, w-corner-1, h-3);
-              }
+            if (resetTransform) {
+                width = Region.clipRound(at.getScaleX() * w);
+                height = Region.clipRound(at.getScaleY() * h);
+                xtranslation = Region.clipRound(at.getScaleX() * x + at.getTranslateX());
+                ytranslation = Region.clipRound(at.getScaleY() * y + at.getTranslateY());
```

Comment:
This probably should be refactored to something like this:
https://github.com/openjdk/jdk/blob/ec4fb47b90c9737dfdc285ebe98367a221c90c79/src/java.desktop/share/classes/sun/java2d/SunGraphics2D.java#L375

Reply:
@mrserb Thank for the suggestion. Are you suggesting that same transformations be applied for width, height and additionally to x&y translations? 
https://github.com/openjdk/jdk/blob/ec4fb47b90c9737dfdc285ebe98367a221c90c79/src/java.desktop/share/classes/sun/java2d/SunGraphics2D.java#L373

Reply:
The methods like Rigion.dimAdd/clipScale should be used everywhere the overflow is possible.

Reply:
Thank you, @mrserb, for your suggestion. I didn't know about those methods.

We already have [JDK-8294680](https://bugs.openjdk.org/browse/JDK-8294680) to refactor this code. If you don't object, I'd rather postpone updating the calculations to a later time. We should absolutely re-use the existing code if it does the same thing.

Reply:
> Thank you, @mrserb, for your suggestion. I didn't know about those methods.
> 
> We already have [JDK-8294680](https://bugs.openjdk.org/browse/JDK-8294680) to refactor this code. If you don't object, I'd rather postpone updating the calculations to a later time. We should absolutely re-use the existing code if it does the same thing.

@aivanov-jdk I think we decided to retain roundHalfDown and make the changes when refactoring the code.

Reply:
> > Thank you, @mrserb, for your suggestion. I didn't know about those methods.
> > We already have [JDK-8294680](https://bugs.openjdk.org/browse/JDK-8294680) to refactor this code. If you don't object, I'd rather postpone updating the calculations to a later time. We should absolutely re-use the existing code if it does the same thing.
> 
> @aivanov-jdk I think we decided to retain roundHalfDown and make the changes when refactoring the code.

The comment above is about `Rigion.dimAdd` and `Rigion.clipScale`.

As for `roundHalfDown`, you had replaced it with `Region.clipRound`. When I was reading the code, I got the impression that you used `Region.clipRound` everywhere but one place. Eventually, I see you've reverted that change, and you're using `roundHalfDown` with no trace of `clipRound` left to be found.

What I saw must have been coming from code comments rather than the real code.

[The comment above](https://github.com/openjdk/jdk/pull/10274#discussion_r990278757) confirms that using `Region.clipRound` gives the same result:

> > The sun.java2d.pipe.Region#clipRound() seems have the same purpose, but it also clip the double to int.
> 
> … It's better to re-use the existing implementation. I tested with it, it gives the same result.

It went into #10681.

I just missed that you had reverted it back.

I guess it's not worth re-doing it once again. We'll handle it in [JDK-8294680](https://bugs.openjdk.org/browse/JDK-8294680) too.

Reply:
Sounds good. I'll have it updated during [JDK-8294680](https://bugs.openjdk.org/browse/JDK-8294680) changes.